{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "72a7fe79",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "from numpy import loadtxt\n",
    "from collections import Counter\n",
    "import matplotlib.colors as mcolors\n",
    "import os\n",
    "import matplotlib\n",
    "matplotlib.use(\"svg\")\n",
    "\n",
    "\n",
    "SIZE = 7\n",
    "\n",
    "def print_map(filename, map_):\n",
    "    df=pd.DataFrame(map_)\n",
    "    df.to_csv(filename+'.csv', index= False, header = False)\n",
    "\n",
    "def extract_latin_name(name):\n",
    "    start = name.find(\"(\")\n",
    "    end = name.find(\")\")\n",
    "    return name[start + 1:end] if start != -1 and end != -1 else name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2d5a36f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#set colormap\n",
    "\n",
    "#colors = ['#000000', '#F3B431', '#CB285B'] #black, yellow, pink\n",
    "colors = ['#dcdcea', '#F2633E', '#5D2265'] #gray, pink, purple\n",
    "#colors = ['#140b34', \"#f6d746\",  '#e55c30'] #dark purple, yellow, orange\n",
    "#colors = ['#F9F6D7', \"#f05e5f\",  '#1a0c3d'] # yellow, pink, purple\n",
    "cmap = mcolors.ListedColormap(colors)\n",
    "bounds = [0, 1, 2, 3]\n",
    "norm = mcolors.BoundaryNorm(bounds, cmap.N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5b34e6b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "filenames = [\n",
    "    \"ALL\",\n",
    "\"Human (Homo sapiens)\",\n",
    "\"Sumatran orangutan (Pongo abelii)\",\n",
    "\"Rhesus monkey (Macaca mlatta)\",\n",
    "\"White-tufted-ear marmoset (Callithrix jacchus)\",\n",
    "\"Gray mouse lemur (Microcebus mrinus)\",\n",
    "\"House mouse (Mus musculus)\",\n",
    "\"Norway rat (Rattus norvegicus)\",\n",
    "\"Guinea pig (Cavia porcellus)\",\n",
    "\"Rabbit (Oryctolagus cuniculus)\",\n",
    "\"Cow (Bos taurus)\",\n",
    "\"Horse (Equus caballus)\",\n",
    "\"Dog (Canis familiaris)\",\n",
    "\"Nine-banded armadillo (Dasypus novemcinctus)\",\n",
    "\"African bush elephant (Loxodonta africana)\",\n",
    "\"Lesser hedgehog tenrec (Echinops telfairi)\",\n",
    "\"Gray short-tailed opossum (Monodelphis domestica)\",\n",
    "\"Tasmanian devil (Sarcophilus harrisii)\",\n",
    "\"Tammar Wallaby (Notamacropus eugenii)\",\n",
    "\"Platypus (Ornithorhynchus anatinus)\",\n",
    "\"Chicken (Gallus gallus)\",\n",
    "\"Rock pigeon (Columba livia)\",\n",
    "\"Zebra finch (Taeniopygia guttata)\",\n",
    "\"American alligator (Alligator mississippiensis)\",\n",
    "\"Western painted turtle (Chrysemys picta bellii)\",\n",
    "\"Green anole lizard (Anolis carolinensis)\",\n",
    "\"Burmese python (Python bvittatus)\",\n",
    "\"Schlegels Japanese gecko (Gekko japonicus)\",\n",
    "\"Tuatara (Sphenodon ptnctatus)\",\n",
    "\"Tropical clawed frog (Xenopus tropicalis)\",\n",
    "\"African clawed frog (Xenopus laevis)\",\n",
    "\"Microcaecilia (Microcaecilia unicolor)\",\n",
    "\"Coelacanth (Latimeria chalumnae)\",\n",
    "\"Asian swamp eel (Monopterus albus)\",\n",
    "\"Pufferfish (Tetraodon nigroviridis)\",\n",
    "\"Cod (Gadus morhua)\",\n",
    "\"Zebrafish (Danio rerio)\",\n",
    "\"Spotted gar (Lepisosteus oculatus)\",\n",
    "\"Australian ghostshark (Callorhinchus milii)\",\n",
    "\"Cloudy Catshark (Scyliorhinus torazame)\",\n",
    "\"Inshore hagfish (Eptatretus burgeri)\",\n",
    "\"Sea Lamprey (Petromyzon marinus)\",\n",
    "\"Sea Squirt (Ciona intestinalis)\",\n",
    "\"Florida lancelet (Branchiostoma floridae)\",\n",
    "\"European lancelet (Branchiostoma lanceolatum)\",\n",
    "\"Saccoglossus (Saccoglossus kowalevskii)\",\n",
    "\"Ptychodera (Ptychodera flava)\",\n",
    "\"Purple sea urchin (Strongylocentrotus purpuratus)\",\n",
    "\"Bat starfish (Patiria miniata)\",\n",
    "\"Fruit fly (Drosophila melanogaster)\",\n",
    "\"Fruit fly (Drosophila simulans)\",\n",
    "\"Fruit fly (Drosophila yakuba)\",\n",
    "\"Fruit fly (Drosophila ananassae)\",\n",
    "\"Fruit fly (Drosophila mojavensis)\",\n",
    "\"Tsetse fly (Glossina pallidipes)\",\n",
    "\"Yellow fever mosquito (Aedes aegypti)\",\n",
    "\"African malaria mosquito (Anopheles gambiae)\",\n",
    "\"Longwing butterfly (Heliconius melpomene)\",\n",
    "\"Red flour beetle (Tribolium castaneum)\",\n",
    "\"Braconid wasp (Diachasmimorpha longicaudata)\",\n",
    "\"Cockroach (Blattella germanica)\",\n",
    "\"Common water flea (Daphnia pulex)\",\n",
    "\"Large common water flea (Daphnia magna)\",\n",
    "\"Deer tick (Ixodes scapularis)\",\n",
    "\"Two-spotted spider mite (Tetranychus urticae)\",\n",
    "\"Arizona bark scorpion (Centruroides sculpturatus)\",\n",
    "\"Atlantic horseshoe crab (Limulus polyphemus)\",\n",
    "\"Roundworm (Caenorhabditis elegans)\",\n",
    "\"Roundworm (Caenorhabditis briggsae)\",\n",
    "\"Large roundworm (Ascaris suum)\",\n",
    "\"Montseny horsehair worm (Gordionus sp.)\",\n",
    "\"Penis worm (Priapulus caudatus)\",\n",
    "\"California two-spot octopus (Octopus bimaculoides)\",\n",
    "\"Common octopus (Octopus vulgaris)\",\n",
    "\"Hawaiian bobtail squid (Euprymna scolopes)\",\n",
    "\"Chambered Nautilus (Nautilus pompilius)\",\n",
    "\"Owl limpet (Lottia gigantea)\",\n",
    "\"Red Abalone (Haliotis rufescens)\",\n",
    "\"Pacific oyster (Magallana gigas)\",\n",
    "\"Japanese carpet shell (Ruditapes philippinarum)\",\n",
    "\"Vernerds tusk shell (Pictodentalium vernedei)\",\n",
    "\"Laevipilina (Laevipilina hyalina)\",\n",
    "\"West Indian fuzzy chiton (Acanthopleura granulata )\",\n",
    "\"Mossy chiton (Mopalia omscosa)\",\n",
    "\"Epimenia (Epimenia babai)\",\n",
    "\"Wirenia (Wirenia argentea)\",\n",
    "\"Polychaete worm (Capitella teleta)\",\n",
    "\"Common brandling worm (Eisenia fetida)\",\n",
    "\"Tubeworm (Owenia fusiformis)\",\n",
    "\"Dumerils clam worm (Platynereis dumerilii)\",\n",
    "\"Peanut worm (Sipunculus nudus)\",\n",
    "\"Dimorphilus gyrociliatus (Dimorphilus grrociliatus)\",\n",
    "\"Lingula (Lingula anatina)\",\n",
    "\"Horseshoe worm (Phoronis australis)\",\n",
    "\"Bootlace worm (Lineus longissimus)\",\n",
    "\"Tiger flatworm (Prostheceraeus crozieri )\",\n",
    "\"Freshwater planarian (Schmidtea mediterranea)\",\n",
    "\"Russian doll killer (Gyrodactylus salaris)\",\n",
    "\"Blood fluke (Schistosoma mansoni)\",\n",
    "\"Hydatid worm (Echinococcus granulosus)\",\n",
    "\"Monogonont Rotifer 1 (Brachionus plicatilis)\",\n",
    "\"Monogonont Rotifer 2 (Brachionus koreanus)\",\n",
    "\"Bdelloid rotifer (Adineta vaga)\",\n",
    "\"Seison (Seison nebaliae)\",\n",
    "\"Thorny-headed worm 1 (Neoechinorhynchus agilis)\",\n",
    "\"Thorny-headed worm 2 (Pomphorhynchus levis)\",\n",
    "\"Xenoturbella (Xenoturbella bocki)\",\n",
    "\"Panther worm (Hofstenia miamia)\",\n",
    "\"Roscoff worm (Symsagittifera roscoffensis)\",\n",
    "\"Starlet sea anemone (Nematostella vectensis)\",\n",
    "\"Deep-sea anemone (Actinernus sp.)\",\n",
    "\"Stony coral (Acropora digitifera)\",\n",
    "\"Freshwater-polyp (Hydra vulgaris)\",\n",
    "\"Amphimedon (Amphimedon queenslandica)\",\n",
    "\"Muellers freshwater sponge (Ephydatia muelleri)\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "befb371c",
   "metadata": {},
   "outputs": [],
   "source": [
    "filenames = [\n",
    "    #invertabrates\n",
    "    ('Caenorhabditis elegans (PRJNA13758) genes (WBcel235)','Roundworm (Caenorhabditis elegans)'),\n",
    "    ('Drosophila melanogaster (Fruit fly) genes (BDGP6.46)','Fruit fly (Drosophila melanogaster)'),\n",
    "    (\"Caenorhabditis briggsae genes\",\"Roundworm (Caenorhabditis briggsae)\"),\n",
    "    (\"Drosophila ananassae genes\",\"Fruit fly (Drosophila ananassae)\"),\n",
    "    (\"Drosophila mojavensis genes\",\"Fruit fly (Drosophila mojavensis)\"),\n",
    "    (\"Drosophila simulans genes\",\"Fruit fly (Drosophila simulans)\"),\n",
    "    (\"Drosophila yakuba genes\",\"Fruit fly (Drosophila yakuba)\"),\n",
    "    (\"Ciona intestinalis genes\",\"Sea Squirt (Ciona intestinalis)\"),\n",
    "    (\"Branchiostoma floridae genes\",\"Florida lancelet (Branchiostoma floridae)\"),\n",
    "    (\"Strongylocentrotus purpuratus genes\",\"Purple sea urchin (Strongylocentrotus purpuratus)\"),\n",
    "    (\"Octopus bimaculoides (California two-spot octopus, UCB-OBI-ISO-001)\",\"California two-spot octopus (Octopus bimaculoides)\"),\n",
    "    #vertabrates\n",
    "    ('Lamprey genes (Pmarinus_7.0)','Sea Lamprey (Petromyzon marinus)'),\n",
    "    ('Hagfish genes (Eburgeri_3.2)','Inshore hagfish (Eptatretus burgeri)'),\n",
    "    ('Elephant shark genes (Callorhinchus_milii-6.1.3)','Australian ghostshark (Callorhinchus milii)'),\n",
    "    ('Spotted gar genes (LepOcu1)','Spotted gar (Lepisosteus oculatus)'),\n",
    "    ('Zebrafish genes (GRCz11)','Zebrafish (Danio rerio)'),\n",
    "    ('Atlantic cod genes (gadMor3.0)','Cod (Gadus morhua)'),\n",
    "    ('Coelacanth genes (LatCha1)','Coelacanth (Latimeria chalumnae)'),\n",
    "    ('Tropical clawed frog genes (Xenopus_tropicalis_v9.1)','Tropical clawed frog (Xenopus tropicalis)'),\n",
    "    ('Tuatara genes (ASM311381v1)','Tuatara (Sphenodon punctatus)'),\n",
    "    ('Green anole genes (AnoCar2.0v2)','Green anole lizard (Anolis carolinensis)'),\n",
    "    ('Painted turtle genes (Chrysemys_picta_bellii-3.0.3)','Western painted turtle (Chrysemys picta bellii)'),\n",
    "    ('Zebra finch genes (bTaeGut1_v1.p)','Zebra finch (Taeniopygia guttata)'),\n",
    "    ('Chicken (maternal Broiler) genes (bGalGal1.mat.broiler.GRCg7b)','Chicken (Gallus gallus)'),\n",
    "    ('Platypus genes (mOrnAna1.p.v1)','Platypus (Ornithorhynchus anatinus)'),\n",
    "    ('Opossum genes (ASM229v1)','Gray short-tailed opossum (Monodelphis domestica)'),\n",
    "    ('Armadillo genes (Dasnov3.0)','Nine-banded armadillo (Dasypus novemcinctus)'),\n",
    "    ('Cow genes (ARS-UCD1.2)','Cow (Bos taurus)'),\n",
    "    ('Dog genes (ROS_Cfam_1.0)','Dog (Canis familiaris)'),\n",
    "    ('Rabbit genes (OryCun2.0)','Rabbit (Oryctolagus cuniculus)'),\n",
    "    ('Guinea Pig genes (Cavpor3.0)','Guinea pig (Cavia porcellus)'),\n",
    "    ('Rat genes (mRatBN7.2)','Norway rat (Rattus norvegicus)'),\n",
    "    ('Mouse genes (GRCm39)','House mouse (Mus musculus)'),\n",
    "    ('Human genes (GRCh38.p13)','Human (Homo sapiens)')\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4ee3672d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "def parse_fas_to_dict(filename):\n",
    "    result = {}\n",
    "    with open(filename, 'r') as f:\n",
    "        label = None\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                label = line[1:4]  # Get first 3 letters after '>'\n",
    "            elif label:\n",
    "                substring = line[1:8]  # Extract str[1:8]\n",
    "                substring = substring.replace('U','T')\n",
    "                if label not in result:\n",
    "                    result[label] = []\n",
    "                result[label].append(substring)\n",
    "    return result\n",
    "\n",
    "def count_unique_sequences(data_dict):\n",
    "    unique_dict = {}\n",
    "    shared_dict = {}\n",
    "\n",
    "    # Flatten all sequences into a key-to-set mapping\n",
    "    key_to_set = {k: set(v) for k, v in data_dict.items()}\n",
    "\n",
    "    for key, seq_set in key_to_set.items():\n",
    "        # Union of all other sets\n",
    "        other_seqs = set().union(*(s for k, s in key_to_set.items() if k != key))\n",
    "        # Sequences that are only in this key\n",
    "        unique_seqs = seq_set - other_seqs\n",
    "        unique_dict[key] = set(unique_seqs)\n",
    "        shared_dict[key] = set(seq_set-unique_seqs)\n",
    "\n",
    "    return unique_dict, shared_dict\n",
    "\n",
    "def update_keys_with_species_names(d, species_filename):\n",
    "    updated_dict = {}\n",
    "    unmatched_keys = []\n",
    "    used_abbrevs = set()\n",
    "    abbrev_to_species = {}\n",
    "    \n",
    "    # Load species lines\n",
    "    with open(species_filename, 'r') as f:\n",
    "        lines = [line.strip() for line in f if line.strip()]\n",
    "    \n",
    "    # Build lookup from abbreviation (like 'hsa') to full species line\n",
    "    for line in lines:\n",
    "        if '(' in line and ')' in line:\n",
    "            try:\n",
    "                inner = line.split('(')[1].split(')')[0]\n",
    "                words = inner.strip().split()\n",
    "                if len(words) >= 2:\n",
    "                    abbrev = words[0][0].lower() + words[1][:2].lower()\n",
    "                    if abbrev in abbrev_to_species:\n",
    "                        raise ValueError(f\"Duplicate species abbreviation found: '{abbrev}' for '{line}'\")\n",
    "                    abbrev_to_species[abbrev] = line\n",
    "            except IndexError:\n",
    "                continue  # Skip malformed lines\n",
    "    \n",
    "    # Replace keys in original dict\n",
    "    for key, value in d.items():\n",
    "        abbrev = key.lower()\n",
    "        if abbrev in abbrev_to_species:\n",
    "            new_key = abbrev_to_species[abbrev]\n",
    "            updated_dict[new_key] = value\n",
    "            used_abbrevs.add(abbrev)\n",
    "        else:\n",
    "            unmatched_keys.append(key)\n",
    "    \n",
    "    # Get unmatched species lines\n",
    "    unmatched_species_lines = [\n",
    "        line for abbr, line in abbrev_to_species.items() if abbr not in used_abbrevs\n",
    "    ]\n",
    "    \n",
    "    return updated_dict, unmatched_keys, unmatched_species_lines\n",
    "\n",
    "# Example usage\n",
    "filename = 'data/ALL-mature.fas'\n",
    "d = parse_fas_to_dict(filename)\n",
    "unique, shared = count_unique_sequences(d)\n",
    "unique_with_labels, unmatched_keys, unmatched_species = update_keys_with_species_names(unique, \"data/names.txt\")\n",
    "shared_with_labels, unmatched_keys, unmatched_species = update_keys_with_species_names(shared, \"data/names.txt\")\n",
    "print(unmatched_keys)\n",
    "print(unmatched_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "861ddd6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_isolated_heatmap_exagg(filename):\n",
    "    #import file\n",
    "    filename_data = loadtxt(\"transcripts/\"+filename+\".txt\", comments=\">\",dtype=\"str\")\n",
    "\n",
    "    #substring from 1, 8\n",
    "    filename_data = [sub[1:8] for sub in filename_data]\n",
    "\n",
    "    #create reverse compliments\n",
    "    \n",
    "    #Replace A with X\n",
    "    filename_data = [sub.replace('A', 'X') for sub in filename_data]\n",
    "    \n",
    "    #Replace T with A\n",
    "    filename_data = [sub.replace('U', 'A') for sub in filename_data]\n",
    "    \n",
    "    #Replace X with T\n",
    "    filename_data = [sub.replace('X', 'T') for sub in filename_data]\n",
    "    \n",
    "    #Replace C with X\n",
    "    filename_data = [sub.replace('C', 'X') for sub in filename_data]\n",
    "    \n",
    "    #Replace G with C\n",
    "    filename_data = [sub.replace('G', 'C') for sub in filename_data]\n",
    "    \n",
    "    #Replace X with G\n",
    "    filename_data = [sub.replace('X', 'G') for sub in filename_data]\n",
    "    \n",
    "    #Reverse all the strings\n",
    "    filename_data = [sub[::-1] for sub in filename_data]\n",
    "    \n",
    "    #create the dictionary\n",
    "    df = pd.read_csv(\"AGTC-7seq.csv\", header = None)\n",
    "    data_arr = np.array(df).flatten()\n",
    "\n",
    "    data_dict = {data_arr[i]: 0 for i in range(0, len(data_arr))}\n",
    "\n",
    "    #Update the dictionary:\n",
    "    for i in filename_data:\n",
    "        data_dict= sequence_count(i,data_dict)\n",
    "\n",
    "    #Convert dictionary back into 1D array\n",
    "    data_arr1 = np.array(list(data_dict.values()))\n",
    "\n",
    "    #Convert 1D array back into a 2D array\n",
    "    data_arr2 = np.reshape(data_arr1, (2**SIZE,2**SIZE))\n",
    "\n",
    "    #save the map as a csv\n",
    "    print_map(\"maps/\"+filename+\"_map\", data_arr2)\n",
    "\n",
    "    # Read the 2D array from the CSV file\n",
    "    df = pd.read_csv(\"maps/\"+filename+\"_map.csv\", header=None)\n",
    "\n",
    "    # Set all values less than 0.5 to be masked\n",
    "    #df = np.ma.masked_where(df < 0.5, df)\n",
    "\n",
    "    # Get the inferno colormap and modify it for masked values\n",
    "    cmap_inferno = mpl.cm.get_cmap(\"inferno\").copy()\n",
    "    cmap_inferno.set_bad(color='black')\n",
    "\n",
    "    # Standardize the data with mean and std, then apply the sigmoid function\n",
    "    def sigmoid(x):\n",
    "        e = np.exp(1)\n",
    "        y = 1 / (1 + e ** (-x))\n",
    "        return y\n",
    "\n",
    "    df = (df - df.mean()) / (df.std())  # Standardize\n",
    "    df = sigmoid(df)  # Apply sigmoid\n",
    "\n",
    "    #DO MAXPOOLING TO MAKE HEATMAPS SMALLER (CAN REMOVE THIS)\n",
    "    #df_maxpool = maximum_filter(df, size=3)\n",
    "\n",
    "    # Create the plot\n",
    "    fig = plt.figure(figsize=(25, 25))\n",
    "    plt.imshow(df, cmap=cmap_inferno)  # Use max-pooled data here\n",
    "    plt.rc('font', size=20)\n",
    "    plt.title(extract_latin_name(filename), fontsize=90)\n",
    "    plt.xticks(ticks=[], labels=[])\n",
    "    plt.yticks(ticks=[], labels=[])\n",
    "    plt.savefig(\"seed_heatmap_raster/\" + filename + \" Map_MaxPool.svg\", bbox_inches='tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3a8f41c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_isolated_heatmap(filename, genome_name, cmap):\n",
    "    #create the dictionary\n",
    "    df = pd.read_csv(\"AGTC-7seq.csv\", header = None)\n",
    "    data_arr = np.array(df).flatten()\n",
    "\n",
    "    data_dict = {data_arr[i]: 0 for i in range(0, len(data_arr))}\n",
    "    \n",
    "    #set shared seeds to 2:\n",
    "    if filename == \"ALL\":\n",
    "        for key in shared_with_labels.keys():\n",
    "            for seed in shared_with_labels[key]:\n",
    "                    data_dict[seed] = 2\n",
    "    else:\n",
    "        if filename in shared_with_labels.keys():\n",
    "            print(\"Found species in shared\")\n",
    "            for seed in shared_with_labels[filename]:\n",
    "                data_dict[seed] = 2\n",
    "    \n",
    "    #set unique seeds to 1\n",
    "    if filename == \"ALL\":\n",
    "        for key in unique_with_labels.keys():\n",
    "            for seed in unique_with_labels[key]:\n",
    "                    data_dict[seed] = 1\n",
    "    else:\n",
    "        if filename in unique_with_labels.keys():\n",
    "            print(\"Found species in unique\")\n",
    "            for seed in unique_with_labels[filename]:\n",
    "                data_dict[seed] = 1\n",
    "        \n",
    "\n",
    "    #Convert dictionary back into 1D array\n",
    "    data_arr1 = np.array(list(data_dict.values()))\n",
    "\n",
    "    #Convert 1D array back into a 2D array\n",
    "    data_arr2 = np.reshape(data_arr1, (2**SIZE,2**SIZE))\n",
    "\n",
    "    #save the map as a csv\n",
    "    print_map(\"maps/\"+filename+\"_map\", data_arr2)\n",
    "\n",
    "    #map 2D array\n",
    "    df = pd.read_csv(\"maps/\"+filename+\"_map.csv\", header = None)\n",
    "\n",
    "    #font\n",
    "    font = {'family' : 'Arial',\n",
    "        'weight' : 'normal',\n",
    "        'size'   : 12}\n",
    "    plt.rc('font', **font)\n",
    "    plt.rcParams['svg.fonttype'] = 'none'\n",
    "    \n",
    "    fig = plt.figure(figsize=(8,8))\n",
    "    plt.imshow(df,cmap=cmap)\n",
    "    plt.rc('font', size=20)\n",
    "    plt.title(filename)\n",
    "    \n",
    "    # labels = ['No Seed', 'Unique Seed', 'Shared Seed']\n",
    "    # cbar = plt.colorbar(ticks=[.35,1,1.65])\n",
    "    # cbar.ax.set_yticklabels(labels)\n",
    "\n",
    "    plt.xticks(ticks=[], labels=[])\n",
    "    plt.yticks(ticks=[], labels=[])\n",
    "    plt.savefig(\"seed_heatmap_raster/\" + genome_name + \".svg\", bbox_inches='tight', format=\"svg\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad3a9457",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found species in shared\n",
      "Found species in unique\n",
      "Found species in shared\n",
      "Found species in unique\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n",
      "/var/folders/wd/wcjpwvbx6ql2slnp2p7cvn2m0000gn/T/com.apple.shortcuts.mac-helper/ipykernel_15444/1465670794.py:62: UserWarning: FigureCanvasSVG is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "for i in filenames:\n",
    "    create_isolated_heatmap(i[1], i[0], cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "dbab72a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "with imageio.get_writer('Seed Heatmap.gif', mode='I') as writer:\n",
    "    for filename in filenames:\n",
    "        image = imageio.v2.imread(\"heatmaps/\"+filename +\" Map.png\")\n",
    "        writer.append_data(image)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f59f30f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is a little calculator to compute the fraction of utilized seeds to unutilized seeds\n",
    "def calc_fraction(filename):\n",
    "    #create the dictionary\n",
    "    df = pd.read_csv(filename+\"_map.csv\", header = None)\n",
    "    data_arr = np.array(df).flatten()\n",
    "    \n",
    "    count_used = 0\n",
    "    \n",
    "    for i in data_arr:\n",
    "        if i != 0:\n",
    "            count_used += 1\n",
    "    ans = count_used / data_arr.size\n",
    "    print(\"Fraction Used: \")\n",
    "    print(ans)\n",
    "    \n",
    "calc_fraction(\"ALL\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
